extensions/amp-date-picker/0.1/amp-date-picker.js

```diff
@@ -228,67 +392,430 @@ class AmpDatePicker extends AMP.BaseElement {
+      this.container_.contains(target) || this.isDateField_(target)
+    );
+
+    if (!clickWasInDatePicker) {
+      this.transitionTo_(DatePickerState.OVERLAY_CLOSED);
+    }
+  }
+
+  /**
+   * Handle focus events in the document to switch between selecting the start
+   * and end dates, and when to open and close the date picker.
+   * @param {!Event} e
+   */
+  handleFocus_(e) {
+    const target = dev().assertElement(e.target);
+    if (this.isDateField_(target)) {
+      if (target == this.startDateField_) {
+        this.updateDateFieldFocus_(this.startDateField_);
+        this.setState_({focusedInput: this.ReactDatesConstants_.START_DATE});
+      } else if (target == this.endDateField_) {
+        this.updateDateFieldFocus_(this.endDateField_);
+        this.setState_({focusedInput: this.ReactDatesConstants_.END_DATE});
+      } else if (target == this.dateField_) {
+        this.updateDateFieldFocus_(this.dateField_);
+      }
+      this.transitionTo_(DatePickerState.OVERLAY_OPEN_INPUT);
+    } else if (this.element.contains(target)) {
+      this.transitionTo_(DatePickerState.OVERLAY_OPEN_PICKER);
+    } else {
+      this.updateDateFieldFocus_(null);
+      this.transitionTo_(DatePickerState.OVERLAY_CLOSED);
+    }
+  }
+
+  /**
+   * For inputs that are valid dates, update the date-picker value.
+   * @param {!Event} e
+   * @private
+   */
+  handleInput_(e) {
+    const target = dev().assertElement(e.target);
+    if (!this.isDateField_(target) || target.type == 'hidden') {
+      return;
+    }
+
+    const property =
+        (target === this.dateField_ ? 'date' :
+          (target === this.startDateField_ ? 'startDate' :
+            (target === this.endDateField_ ? 'endDate' : '')));
+    const moment = this.createMoment_(target.value);
+    if (moment.isValid()) {
+      this.setState_({[property]: moment});
+    }
+  }
+
+  /**
+   * Handle changing values of the attached input fields, and the hotkey for
+   * closing the date picker.
+   * @param {!Event} e
+   */
+  handleKeydown_(e) {
+    const target = dev().assertElement(e.target);
+    if (this.isDateField_(target)) {
+      this.handleInputKeydown_(e);
+    } else {
+      this.handleDocumentKeydown_(e);
+    }
+  }
+
+  /**
+   * Close the date picker overlay when the escape key is pressed.
+   * @param {!Event} e
+   */
+  handleDocumentKeydown_(e) {
+    if (e.keyCode == KeyCodes.ESCAPE &&
+        this.mode_ == DatePickerMode.OVERLAY &&
+        this.element.contains(this.document_.activeElement)) {
+      this.transitionTo_(DatePickerState.OVERLAY_CLOSED);
+    }
+    return true;
+  }
+
+  /**
+   * Handle the states for keyboard input.
+   * @param {!Event} e
    * @private
    */
-  parseSrcTemplates_(srcTemplatePromise) {
-    if (!srcTemplatePromise) {
-      return Promise.resolve();
+  handleInputKeydown_(e) {
+    const target = dev().assertElement(e.target);
+    if (!this.isDateField_(target) || target.type == 'hidden') {
```

Comment:
move the `type == hidden` to `handleKeydown_`. There shouldn't be a need to check `!this.isDateField_(target)` given `handleKeydown_` does that already.